agent/src/main/java/com/microsoft/hydralab/agent/runner/TestRunner.java

```diff
@@ -139,18 +143,32 @@ private void presetForMicrosoftLauncherApp(DeviceInfo deviceInfo, TestTask testT
 
 
     protected void reInstallApp(DeviceInfo deviceInfo, TestTask testTask, Logger reportLogger) throws Exception {
-        if (testTask.getRequireReinstall()) {
+        if (testTask.getRequireReinstall() || deviceManager instanceof IOSDeviceManager) {
             deviceManager.uninstallApp(deviceInfo, testTask.getPkgName(), reportLogger);
-            deviceManager.uninstallApp(deviceInfo, testTask.getTestPkgName(), reportLogger);
             ThreadUtils.safeSleep(1000);
         } else if (testTask.getRequireClearData()) {
             deviceManager.resetPackage(deviceInfo, testTask.getPkgName(), reportLogger);
         }
         checkTestTaskCancel(testTask);
 
         deviceManager.installApp(deviceInfo, testTask.getAppFile().getAbsolutePath(), reportLogger);
-        deviceManager.installApp(deviceInfo, testTask.getTestAppFile().getAbsolutePath(), reportLogger);
+    }
 
+    protected void reInstallTestApp(DeviceInfo deviceInfo, TestTask testTask, Logger reportLogger) throws Exception {
+        if (testTask.getTestAppFile() == null) {
+            return;
+        }
+        if (StringUtils.isEmpty(testTask.getTestPkgName())) {
+            return;
+        }
+        if (!testTask.getTestAppFile().exists()) {
```

Comment:
Currently, the testAppFile could be .apk file (Espresso) or .jar file(Appium). So, we could not install it directly.

Reply:
Will this work?
![image](https://user-images.githubusercontent.com/8344245/208370169-c3f2ebd4-b310-4ed5-b2f4-4cf0d5df071c.png)